walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,90 @@
+  *  @return A list of open meeting timeslots.  
+  */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+      List<Event> eventsList = new ArrayList<>(events);
+      sort(eventsList, new Comparator<Event>() {
+        public int compare (Event e1, Event e2) {
+            return TimeRange.ORDER_BY_START.compare(e1.getWhen(), e2.getWhen());
+          }
+        });
+      /** We want to remove any attendees not attending or events less-than/equal to 0. */
+      eventsList.removeIf(e -> (
+          e.getWhen().duration() <= 0 || disjoint(e.getAttendees(), request.getAttendees())));
+    
+      if( request.getAttendees().isEmpty()){
+          return asList(TimeRange.WHOLE_DAY);
+      }
+      if(request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+          return EMPTY_LIST;
+      }
+      if(eventsList.size()==0){
+          return asList(TimeRange.WHOLE_DAY);
+      }
+      if(eventsList.size()==1){
+          return asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, eventsList.get(0).getWhen().start(), false),
+          TimeRange.fromStartEnd(eventsList.get(0).getWhen().end(), TimeRange.END_OF_DAY, true)); 
+       }
+
+      return getOpenTimeSlots(eventsList, request);
+  }
+  /**
+  *  This method gets the open time slots by looping through a list of events
+  *  given that the list is sorted by start time. 
+  *  @param eventsList list of events in sorted order (by start time).
+  *  @param request he duration of requested meeting and people attending.
+  *  @return A list of open meeting timeslots. 
+  */
+  private Collection<TimeRange> getOpenTimeSlots(List<Event> eventsList, MeetingRequest request){
+      Collection<TimeRange> queryResult = new ArrayList<>();
+      for(int i=0;i<eventsList.size();i++){
+        /** 
+        *  First event in the list, check if there is enough time between event start and the start of the day. 
+        *  If there is enough time, add to the result. Since the events are sorted by start time we can check the
+        *  first index.
+        */
+        if(i==0){
+            if(eventsList.get(i).getWhen().start()-request.getDuration()>=TimeRange.START_OF_DAY){
+              queryResult.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY,eventsList.get(i).getWhen().start(), false));
+            }
+        }
+        if(i != eventsList.size()-1){
+          if(eventsList.get(i).getWhen().overlaps(eventsList.get(i+1).getWhen())){
+            /** Check for overlap.  Don't need to check for the reverse, since it is pre-sorted. */
+            if(eventsList.get(i).getWhen().contains(eventsList.get(i+1).getWhen())){
+                /** 
+                *  If one event contains another event. Set i+1 to i.  
+                *  This is to get the earliest start and latest end time for an event.
+                */
+                 eventsList.set(i+1, eventsList.get(i));
```

Comment:
hmmm are you mutating the eventList input list?   I would buy the book Clean Code which has a section that say methods should not have any side effects.   (And if it does, clearly document the side effects.)

Reply:
In a future CL I copy eventsList into another list.  So there will be no side effects.